content/blog/2023/2023-01-25-junit5-parameterized-tests.md

```diff
@@ -0,0 +1,460 @@
+
+```java
+@ParameterizedTest
+@ValueSource(ints = { 100 })
+void checkExplicitArgumentConversion(@ConvertWith(StringSimpleArgumentConverter.class) String argument) {
+	assertEquals("100", argument);
+}
+
+public class StringSimpleArgumentConverter extends SimpleArgumentConverter {
+
+	@Override
+	protected Object convert(Object source, Class<?> targetType) throws ArgumentConversionException {
+		return String.valueOf(source);
+	}
+}
+```
+
+### Argument Aggregation
+
+#### @ArgumentsAccessor
+
+By default, each argument provided to a `@ParameterizedTest` method corresponds to a `single method parameter`. Due to this, when argument sources that supply a large number of arguments can lead to large method signatures.
+To solve this problem, we can use `ArgumentsAccessor` instead of declaring multiple parameters. The type conversion is supported as discussed in Implicit conversion above.
+
+```java
+@ParameterizedTest
+@CsvSource({ "John, 20", 
+		     "Harry, 30" })
+void checkArgumentsAccessor(ArgumentsAccessor arguments) {
+	Person person = new Person(arguments.getString(0), arguments.getInteger(1));
+	assertTrue(person.getAge() > 19, person.getName() + " is a teenager");
+}
+```
+
+#### Custom Aggregators
+
+We saw using an `ArgumentsAccessor` can access the @ParameterizedTest method’s arguments directly. What if we want to declare the same ArgumentsAccessor in multiple tests? JUnit5 solves this by providing custom, reusable aggregators.
+
+- **@AggregateWith** - Implement the `ArgumentsAggregator` interface and register it via the `@AggregateWith` annotation in the @ParameterizedTest method. When we execute the test, it provides the aggregation result as an argument for the corresponding test. The implementation of ArgumentsAggregator can be an external class or a static nested class.
+
+```java
+@ParameterizedTest
+@CsvSource({ "John, 20", 
+			 "Harry, 30" })
+void checkArgumentsAggregator(@AggregateWith(PersonArgumentsAggregator.class) Person person) {
+	assertTrue(person.getAge() > 19, person.getName() + " is a teenager");
+}
+
+public class PersonArgumentsAggregator implements ArgumentsAggregator {
+
+	@Override
+	public Object aggregateArguments(ArgumentsAccessor arguments, ParameterContext context)
+			throws ArgumentsAggregationException {
+		return new Person(arguments.getString(0), arguments.getInteger(1));
+	}
+}
+```
+
+- **Custom Annotation** - When we see we are repeating `@AggregateWith(PersonArgumentsAggregator.class)` declaration for each test, we can create a custom annotation such as `@CsvToPerson` that can be meta-annotated with `@AggregateWith(PersonArgumentsAggregator.class)` to have a similar effect.
```

Comment:
This solution is useful but it is not directly related to JUnit 5's features. It would be better suited for an article specifically focused on custom annotations. We cann remove this section from the current article to stay on topic and provide a clear focus on JUnit 5's features.

Reply:
Removed this.